### DISCUSSION TITLE
Qualitative evaluation of agents

### DISCUSSION THREAD

**Discussion by jbirkesteen** [RasmusBrostroem/ConnectFourRL#69] [Ideas]
_To discuss how we can evaluate the strategies deployed by agents._
At the moment, the logging of several quantitative metrics is implemented with neptune, allowing us to assess the progress and performance of agents during their training phase.  
However, we need other approaches in order to understand **how** they play. As a starting point this involves watching agents play and judging which kinds of strategies they are able to use.  

**Points to discuss:**
1. How can we enhance the play-watching experience to further the understanding of what is going on?  
1. Can we come up with some way of extending this analysis, e.g. in terms of statistics to track and/or visualise?

**Comment by jbirkesteen** [comment]
### Separate module for evaluation/testing
How do we best incorporate the enhancements in this discussion in our project? Do we want to add a separate module/part of the project which contains everything related to evaluation?  

_(this might not be relevant right now, but we should keep it in the back of our heads when we start implementing stuff in this discussion)_

**Comment by jbirkesteen** [comment]
### Visualise probabilities
Display the output of the network (column probabilities) when it makes a move. Highlight the one chosen (for instance, by writing it in italic/bold). 
These probabilities could be displayed either above the board (where we've previously shown winner/loser/tie status) or below the board.

**Comment by jbirkesteen** [comment]
### Click button for next move
Currently, when the game is displayed, the pause before moves is equal for the entire game.  
However, game states are typically simple and easy to understand in the early phases of the game, while they get increasingly complex later on. This is not ideal for analysing strategies.
**Let the user watch the game in their own pace. Wait for user input 'N' before progressing to the Next move.**

**Comment by jbirkesteen** [comment]
### Find sources on strategies
Typically, when RL-researchers try to win new games, they work closely together with experts on the game in order to understand their agent's flaws better (for instance, deepmind did this for both Go and Stratego).  
Can we find some sources to improve our own understanding of the game to better know what to look out for?

This might also give us some ideas down the road for better visualisations in the analysis.

**Comment by jbirkesteen** [comment]
### Which color is my agent?
Currently, one can not tell from the display of the game which agent is red/yellow.  
This is based on the value of `player_piece` in the initialisation of agents, but if we can find some clear way of showing it directly in the display, I think that would be an enhancement.

**Comment by jbirkesteen** [comment]
### Highlight mistakes
Highlight/visualise when a player/agent misses a game-deciding move. This might make it easier to convey our reasoning about the capabilities of our agents, and could be very beneficial in the mid/late phases of the game when the game states become increasingly complex.

**Reply by RasmusBrostroem** [reply]
Yes, and we could log to neptune how many times a player missed a winning move, or didn't block a field that the opponent could win in.
